From 518442282ceee4dfbe4b4c4f0f914173c6fab310 Mon Sep 17 00:00:00 2001
From: Sridharan S N <quic_sridsn@quicinc.com>
Date: Thu, 7 Mar 2024 13:27:51 +0530
Subject: [PATCH] drivers: remoteproc: qcom_q6v5: Complete pending waits on
 rproc crash

Rproc start/stop/spawn might be waiting for completion from IRQ when a
rproc crashes. In such cases, the recovery waits for the timeout before
proceeding as mutex is held by the boot/shutdown threads.

Complete all pending waits for an rproc when fatal or wdog IRQ is
received. In the wait_for_completion function, check if the
corresponding ack flag is set and return success/failure. This will
be set only if completion is from the IRQ and not from crash.

Change-Id: I8ed32bb626fbc6f582b4639806bb64574926954a
Signed-off-by: Vignesh Viswanathan <quic_viswanat@quicinc.com>
Signed-off-by: Sridharan S N <quic_sridsn@quicinc.com>
---
 drivers/remoteproc/qcom_q6v5.c | 41 +++++++++++++++++++++++++++++-----
 drivers/remoteproc/qcom_q6v5.h |  4 ++++
 2 files changed, 39 insertions(+), 6 deletions(-)

Index: linux-6.6.116/drivers/remoteproc/qcom_q6v5.c
===================================================================
--- linux-6.6.116.orig/drivers/remoteproc/qcom_q6v5.c	2026-01-27 00:41:21.378973273 +0530
+++ linux-6.6.116/drivers/remoteproc/qcom_q6v5.c	2026-01-27 00:44:22.940319288 +0530
@@ -60,9 +60,13 @@
 
 	reinit_completion(&q6v5->start_done);
 	reinit_completion(&q6v5->stop_done);
+	reinit_completion(&q6v5->spawn_done);
 
 	q6v5->running = true;
 	q6v5->handover_issued = false;
+	q6v5->start_ack = false;
+	q6v5->stop_ack = false;
+	q6v5->spawn_ack = false;
 
 	enable_irq(q6v5->handover_irq);
 
@@ -96,6 +100,7 @@
 
 	/* Sometimes the stop triggers a watchdog rather than a stop-ack */
 	if (!q6v5->running) {
+		q6v5->stop_ack = true;
 		complete(&q6v5->stop_done);
 		return IRQ_HANDLED;
 	}
@@ -107,6 +112,11 @@
 		dev_err(q6v5->dev, "watchdog without message\n");
 
 	q6v5->running = false;
+
+	/* Complete any pending waits for this rproc */
+	complete(&q6v5->spawn_done);
+	complete(&q6v5->start_done);
+	complete(&q6v5->stop_done);
 	rproc_report_crash(q6v5->rproc, RPROC_WATCHDOG);
 
 	return IRQ_HANDLED;
@@ -128,6 +138,11 @@
 		dev_err(q6v5->dev, "fatal error without message\n");
 
 	q6v5->running = false;
+
+	/* Complete any pending waits for this rproc */
+	complete(&q6v5->spawn_done);
+	complete(&q6v5->start_done);
+	complete(&q6v5->stop_done);
 	rproc_report_crash(q6v5->rproc, RPROC_FATAL_ERROR);
 
 	return IRQ_HANDLED;
@@ -137,6 +152,7 @@
 {
 	struct qcom_q6v5 *q6v5 = data;
 
+	q6v5->start_ack = true;
 	complete(&q6v5->start_done);
 
 	return IRQ_HANDLED;
@@ -156,7 +172,12 @@
 	int ret;
 
 	ret = wait_for_completion_timeout(&q6v5->start_done, timeout);
-	return !ret ? -ETIMEDOUT : 0;
+	if (!ret) {
+		disable_irq(q6v5->handover_irq);
+		return -ETIMEDOUT;
+	} else {
+		return q6v5->start_ack ? 0 : -ERESTARTSYS;
+	}
 }
 EXPORT_SYMBOL_GPL(qcom_q6v5_wait_for_start);
 
@@ -178,6 +199,7 @@
 {
 	struct qcom_q6v5 *q6v5 = data;
 
+	q6v5->spawn_ack = true;
 	complete(&q6v5->spawn_done);
 
 	return IRQ_HANDLED;
@@ -187,6 +209,7 @@
 {
 	struct qcom_q6v5 *q6v5 = data;
 
+	q6v5->stop_ack = true;
 	complete(&q6v5->stop_done);
 
 	return IRQ_HANDLED;
@@ -204,7 +227,7 @@
 	int ret;
 
 	q6v5->running = false;
-
+	q6v5->stop_ack = false;
 	/* Don't perform SMP2P dance if remote isn't running */
 	if (q6v5->rproc->state != RPROC_RUNNING || qcom_sysmon_shutdown_acked(sysmon))
 		return 0;
@@ -216,7 +239,10 @@
 
 	qcom_smem_state_update_bits(q6v5->state, BIT(q6v5->stop_bit), 0);
 
-	return ret == 0 ? -ETIMEDOUT : 0;
+	if (!ret)
+		return -ETIMEDOUT;
+	else
+		return q6v5->stop_ack ? 0 : -ERESTARTSYS;
 }
 EXPORT_SYMBOL_GPL(qcom_q6v5_request_stop);
 
@@ -230,6 +256,7 @@
 {
 	int ret;
 
+	q6v5->spawn_ack = false;
 	ret = qcom_smem_state_update_bits(q6v5->spawn_state,
 					  BIT(q6v5->spawn_bit), BIT(q6v5->spawn_bit));
 
@@ -238,7 +265,10 @@
 	qcom_smem_state_update_bits(q6v5->spawn_state,
 				    BIT(q6v5->spawn_bit), 0);
 
-	return ret == 0 ? -ETIMEDOUT : 0;
+	if (!ret)
+		return -ETIMEDOUT;
+	else
+		return q6v5->spawn_ack ? 0 : -ERESTARTSYS;
 }
 EXPORT_SYMBOL_GPL(qcom_q6v5_request_spawn);
 
@@ -286,6 +316,7 @@
 
 	init_completion(&q6v5->start_done);
 	init_completion(&q6v5->stop_done);
+	init_completion(&q6v5->spawn_done);
 
 	q6v5->wdog_irq = platform_get_irq_byname(pdev, "wdog");
 	if (q6v5->wdog_irq < 0)
Index: linux-6.6.116/drivers/remoteproc/qcom_q6v5.h
===================================================================
--- linux-6.6.116.orig/drivers/remoteproc/qcom_q6v5.h	2026-01-27 00:41:21.378973273 +0530
+++ linux-6.6.116/drivers/remoteproc/qcom_q6v5.h	2026-01-27 00:41:21.378973273 +0530
@@ -40,6 +40,10 @@
 	struct completion stop_done;
 	struct completion spawn_done;
 
+	bool start_ack;
+	bool stop_ack;
+	bool spawn_ack;
+
 	int remote_id;
 	int crash_reason;
 
